text/0225-configuration.md

```diff
@@ -0,0 +1,335 @@
+
+* file doesn't exist or is invalid
+* configuration parsed is invalid
+
+#### Python Parse example
+
+```python
+
+filepath = "./config.yaml"
+
+
+try:
+  cfg = opentelemetry.Parse(filepath)
+except Exception as e:
+  print(e)
+
+filepath = "./config.json"
+
+try:
+  cfg = opentelemetry.Parse(filepath)
+except Exception as e:
+  raise e
+
+```
+
+#### Go Parse example
+
+```go
+
+filepath := "./config.yaml"
+cfg, err := otel.Parse(filepath)
+if err != nil {
+  return err
+}
+
+filepath := "./config.json"
+cfg, err := otel.Parse(filepath)
+if err != nil {
+  return err
+}
+
+```
+
+Implementations *MUST* allow users to specify an environment variable to set the configuration file. This gives flexibility to end users of implementations that do not support command line arguments. The proposed name for this variable:
+
+* `OTEL_CONFIG_FILE`
+
+The format for the configuration file will be detected using the file extension of this variable.
+
+### Configuration model
+
+To allow SDKs and instrumentation libraries to accept configuration without having to implement the parsing logic, a `Configuration` model *MAY* be provided by implementations. This object:
+
+* has already been parsed from a file or data structure
+* is structurally valid (errors may yet occur when SDK or instrumentation interprets the object)
+
+### Configuration file
+
+The following demonstrates an example of a configuration file format (full example [here](https://github.com/MrAlias/otel-schema/blob/main/config.yaml)):
+
+```yaml
+# include version specification in configuration files to help with parsing and schema evolution.
+file_format: 0.1
+sdk:
+  # Disable the SDK for all signals.
+  #
+  # Boolean value. If "true", a no-op SDK implementation will be used for all telemetry
+  # signals. Any other value or absence of the variable will have no effect and the SDK
+  # will remain enabled. This setting has no effect on propagators configured through
+  # the OTEL_PROPAGATORS variable.
+  #
+  # Environment variable: OTEL_SDK_DISABLED
+  disabled: false
+  # Configure resource attributes and resource detection for all signals.
+  resource:
+    # Key-value pairs to be used as resource attributes.
+    #
+    # Environment variable: OTEL_RESOURCE_ATTRIBUTES
+    attributes:
+      # Sets the value of the `service.name` resource attribute
+      #
+      # Environment variable: OTEL_SERVICE_NAME
+      service.name: !!str "unknown_service"
+  # Configure context propagators. Each propagator has a name and args used to configure it. None of the propagators here have configurable options so args is not demonstrated.
+  #
+  # Environment variable: OTEL_PROPAGATORS
+  propagators: [tracecontext, baggage]
+  # Configure the tracer provider.
+  tracer_provider:
+    # Span exporters. Each exporter key refers to the type of the exporter. Values configure the exporter. Exporters must be associated with a span processor.
+    exporters:
```

Comment:
How would I be able to support an exporter that is not built-in to the tracing spec? For example, how would I be able to configure the `google-trace` exporter [1]? Currently, OpenTelemetry specification appears to only favour one commercial tracing backend.

[1] https://github.com/GoogleCloudPlatform/opentelemetry-operations-js#opentelemetry-google-cloud-trace-exporter

Reply:
Would it be beneficial to allow to `register` exporters/propagators etc to the SDK? So a user would be able to register `google-trace` and then at the same time be able to use `OTEL_PROPAGATORS` and similar variables?

Reply:
Yup this proposal intends custom extension components (samplers, exporters, propagators, processors) to be referenced in the config file. How exactly those extension components are resolved is a language specific problem. In java, we use SPIs (service provider interface) to allow custom components to be provided by libraries or users.